MantidPlot/src/PlotCurve.cpp

```diff
@@ -278,75 +297,109 @@ void PlotCurve::computeWaterfallOffsets()
   d_x_offset = 0.0;
   d_y_offset = 0.0;
 
-  if (g->isWaterfallPlot()){
+  if (g->isWaterfallPlot()) {
     int index = g->curveIndex(this);
     int curves = g->curves();
     auto firstCurve = g->curve(0);
     // Get the minimum value of the first curve in this plot
     double ymin = firstCurve ? firstCurve->minYValue() : 0.0;
-    PlotCurve *c = dynamic_cast<PlotCurve*>(g->curve(0));
-    if (index > 0 && c){
-      d_x_offset = index*g->waterfallXOffset()*0.01*plot->canvas()->width()/(double)(curves - 1);
-      d_y_offset = index*g->waterfallYOffset()*0.01*plot->canvas()->height()/(double)(curves - 1);
+    PlotCurve *c = dynamic_cast<PlotCurve *>(g->curve(0));
+    if (index > 0 && c) {
+      // Compute offsets based on the maximum value for the curve
+      double x_offset = index * g->waterfallXOffset() * 0.01 *
+                        g->curve(0)->maxXValue() / (double)(curves - 1);
+      double y_offset = index * g->waterfallYOffset() * 0.01 *
+                        g->curve(0)->maxYValue() / (double)(curves - 1);
+
+      // Clone the workspace so that the data is not corrupted when applying
+      // offsets
+      std::unique_ptr<MatrixWorkspace> wksp_clone =
```

Comment:
I don't think this is the best approach to the problem. You are correct that we don't want to affect the original data but if I have a large dataset, plot a few curves then display a waterfall plot I wouldn't expect my memory requirements to double.